<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using the `sager` package to import and analyse `sage` results in R • sager</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using the `sager` package to import and analyse `sage` results in R">
<meta property="og:description" content="sager">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sager</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/sager.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/UCLouvain-CBIO/sager/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using the <code>sager</code> package to import
and analyse <code>sage</code> results in R</h1>
                        <h4 data-toc-skip class="author">Laurent
Gatto</h4>
            <address class="author_afil">
      Computational Biology and Bioinformatics,
UCLouvain<br><a class="author_email" href="mailto:#"></a><a href="mailto:laurent.gatto@uclouvain.be" class="email">laurent.gatto@uclouvain.be</a>
      </address>
                  
            <h4 data-toc-skip class="date">9 November 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/UCLouvain-CBIO/sager/blob/HEAD/vignettes/sager.Rmd" class="external-link"><code>vignettes/sager.Rmd</code></a></small>
      <div class="hidden name"><code>sager.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="section level3">
<h3 id="the-sager-package">The <code>sager</code> package<a class="anchor" aria-label="anchor" href="#the-sager-package"></a>
</h3>
<p><a href="https://lazear.github.io/sage/" class="external-link">Sage</a> is a new
cross-platform, extremely performant, open source proteomics search
engine <a href="https://github.com/lazear" class="external-link">Michael Lazear</a>. It uses
mzML files as input, is parameterised by a <a href="https://github.com/lazear/sage/blob/master/DOCS.md" class="external-link">json file</a>
and is executed from the command line. In addition to peptide
identification, sage also includes a variety of advanced features such
as retention time prediction and quantification (both isobaric &amp;
LFQ). It produces two files: <code>results.sage.tsv</code> with the
identification results and, when configured to quantify features,
<code>quant.tsv</code>.</p>
<p>The <code>sager</code> package uses these two files as input. It
facilitates their import into established Bioconductor classes:</p>
<ul>
<li><p>Identification results are parsed and imported as <a href="https://rformassspectrometry.github.io/PSMatch/" class="external-link">PSMatch::PSM()</a>
objects with <code><a href="../reference/sagePSM.html">sagePSM()</a></code>.</p></li>
<li><p>Quantitation (and identification) results are parsed, merged and
imported as <a href="https://rformassspectrometry.github.io/QFeatures/articles/QFeatures.html" class="external-link">QFeatures::QFeatures()</a>
objects with <code><a href="../reference/sageQFeatures.html">sageQFeatures()</a></code>.</p></li>
<li><p>Sage results can also be integrated with a <a href="https://rformassspectrometry.github.io/Spectra/" class="external-link">Spectra::Spectra()</a>
instance (holding the raw data) into a <a href="https://rformassspectrometry.github.io/MsExperiment/" class="external-link">MsExperiment::MsExperiment()</a>
object, as documented in this document.</p></li>
</ul>
<p>The goal of the <code>sager</code> package can thus be summarised as
a way to leverage the existing <a href="https://www.rformassspectrometry.org/" class="external-link">R for Mass Spectrometry</a>
and <a href="https://bioconductor.org/" class="external-link">Bioconductor</a> infrastructure
to analyse sage results.</p>
</div>
<div class="section level3">
<h3 id="installation-instructions">Installation instructions<a class="anchor" aria-label="anchor" href="#installation-instructions"></a>
</h3>
<p>To install <code>sager</code> and its dependencies, you’ll need the
<code>BiocManager</code> package, that can be installed from CRAN
(unless you already have it):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre></div>
<p>Note that <code>sager</code> is under constant development, and thus
will depend on some under-development dependencies, that haven’t made it
into the stable releases yet. Here are the instructions to install
these:</p>
<p>To install the <em>devel</em> version of <code>PSMatch</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"RforMassSpectrometry/PSMatch"</span><span class="op">)</span></span></code></pre></div>
<p>To install <code>sager</code> and its dependencies, you can run the
following command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"UCLouvain-CBIO/sager"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="analysing-sage-results">Analysing <code>sage</code> results<a class="anchor" aria-label="anchor" href="#analysing-sage-results"></a>
</h2>
<div class="section level3">
<h3 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">sager</span><span class="op">)</span></span></code></pre></div>
<p>The <code>sager</code> package provides example identification and
quantitation (procuced by <code>sage</code>) and the raw data
(<code>mzML</code> files) that were used the generate the former.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sagerData.html">sagerAvailableData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; quant    id  mzml </span></span>
<span><span class="co">#&gt;  TRUE  TRUE  TRUE</span></span></code></pre></div>
<p>Data can be added with the <code><a href="../reference/sagerData.html">sagerAddData()</a></code> function, that
avoid downloading data files if they are already available in the
package’s local cache directory:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sagerData.html">sagerAddData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'quant' already available.</span></span>
<span><span class="co">#&gt; 'id' already available.</span></span>
<span><span class="co">#&gt; 'mzml' already available.</span></span></code></pre></div>
<p>For details, including updating local data and their provenance,
please see the <code><a href="../reference/sagerData.html">?sagerData</a></code> manual page.</p>
</div>
<div class="section level3">
<h3 id="identification-data">Identification data<a class="anchor" aria-label="anchor" href="#identification-data"></a>
</h3>
<p>As noted above, identification data is handled by the <a href="https://rformassspectrometry.github.io/PSMatch/" class="external-link">PSMatch</a>
package. This section only provides a very short example of how to use
it. For more details, see the package’s <a href="https://rformassspectrometry.github.io/PSMatch/" class="external-link">website and
vignettes</a>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/PSM" class="external-link">PSMatch</a></span><span class="op">)</span></span>
<span><span class="va">psm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sagePSM.html">sagePSM</a></span><span class="op">(</span><span class="fu"><a href="../reference/sagerData.html">sagerIdData</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">psm</span></span>
<span><span class="co">#&gt; PSM with 977 rows and 34 columns.</span></span>
<span><span class="co">#&gt; Spectra: 595 unique</span></span>
<span><span class="co">#&gt;   db: 679 target, 298 decoy</span></span>
<span><span class="co">#&gt;   ranks: 1:977 </span></span>
<span><span class="co">#&gt; Peptides: 976 unique, 0 multiple</span></span>
<span><span class="co">#&gt; Proteins: 947 </span></span>
<span><span class="co">#&gt; names(34): peptide proteins ... protein_fdr ms1_intensity</span></span></code></pre></div>
<p>Here we briefly visualise the hyperscore and spectrum FDR densities
for the forward and decoy/reverse hits:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">psm</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">hyperscore</span>,</span>
<span>               colour <span class="op">=</span> <span class="va">label</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="sager_files/figure-html/ggplot-1.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">psm</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">spectrum_fdr</span>,</span>
<span>               colour <span class="op">=</span> <span class="va">label</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="sager_files/figure-html/ggplog2-1.png" width="700"></p>
<p>Below, we filter the PSMs removing decoy hits, PSMs of rank &gt; 1
(none in the small testing data), shared peptides (none in the small
testing data) and those that have a spectrum FDR below 0.05:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PSMatch/man/filterPSMs.html" class="external-link">filterPSMs</a></span><span class="op">(</span><span class="va">psm</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/PSMatch/man/filterPSMs.html" class="external-link">filterPsmFdr</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Starting with 977 PSMs:</span></span>
<span><span class="co">#&gt; Removed 298 decoy hits.</span></span>
<span><span class="co">#&gt; Removed 0 PSMs with rank &gt; 1.</span></span>
<span><span class="co">#&gt; Removed 0 shared peptides.</span></span>
<span><span class="co">#&gt; 679 PSMs left.</span></span>
<span><span class="co">#&gt; Removed 370 PSMs with FDR &lt; spectrum_fdr.</span></span></code></pre></div>
<p>This leaves us with 309 PSMs matched against 294 scans.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psm</span></span>
<span><span class="co">#&gt; PSM with 309 rows and 34 columns.</span></span>
<span><span class="co">#&gt; Spectra: 294 unique</span></span>
<span><span class="co">#&gt;   db: 309 target, 0 decoy</span></span>
<span><span class="co">#&gt;   ranks: 1:309 </span></span>
<span><span class="co">#&gt; Peptides: 308 unique, 0 multiple</span></span>
<span><span class="co">#&gt; Proteins: 297 </span></span>
<span><span class="co">#&gt; names(34): peptide proteins ... protein_fdr ms1_intensity</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="quantitative-data">Quantitative data<a class="anchor" aria-label="anchor" href="#quantitative-data"></a>
</h3>
<p>As noted above, quantitation data is handled by the <a href="https://rformassspectrometry.github.io/QFeatures/" class="external-link">QFeatures</a>
package. This section only provides a very short example of how to use
it. For more details, see the package’s <a href="https://rformassspectrometry.github.io/QFeatures/" class="external-link">website and
vignettes</a>.</p>
<p>We will also make use of some function from the <a href="https://github.com/UCLouvain-CBIO/scp" class="external-link">scp</a> package.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/QFeatures" class="external-link">QFeatures</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://UCLouvain-CBIO.github.io/scp" class="external-link">scp</a></span><span class="op">)</span></span></code></pre></div>
<p>The quantitation data has been produced from running
<code>sage</code> on three raw data files. These three files can
correspond to two different experimental designs:</p>
<ol style="list-style-type: decimal">
<li><p>The three acquisitions correspond to three fractions of the same
set of 11 multiplexed samples. In this case, that we’ll call the
<em>Multiple fractions for the same set of samples</em>, we end up with
a single quantitation matrix with 11 columns and row containing the
features from all files.</p></li>
<li><p>The three acquisitions correspond to single acquisitions from 3
different sets of 11 samples. In this case, that we’ll call the
<em>Different sets of samples per acquisition</em>, we end up with three
quantitation matrices, each with 11 columns and rows containing the
features from the respective files.</p></li>
</ol>
<p>Both sitations are illustrated below using the
<code><a href="../reference/sageQFeatures.html">sageQFeatures()</a></code> function , and a differenciated by the
value of the <code>splitBy</code> argument. Note that below, the
<code>byQuant</code> argument is set to accomodate the old sage default
header; this <code>byQuant</code> argument wouldn’t need to be set
explicitly for newer output files.</p>
<div class="section level4">
<h4 id="multiple-fractions-for-the-same-set-of-samples">Multiple fractions for the same set of samples<a class="anchor" aria-label="anchor" href="#multiple-fractions-for-the-same-set-of-samples"></a>
</h4>
<p>Let’s read the data in, specifying that the tables should not be
split into different quantitative assays.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sageQFeatures.html">sageQFeatures</a></span><span class="op">(</span><span class="fu"><a href="../reference/sagerData.html">sagerQuantData</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                     <span class="fu"><a href="../reference/sagerData.html">sagerIdData</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                     byQuant <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"file"</span>, <span class="st">"scannr"</span><span class="op">)</span>,</span>
<span>                     splitBy <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">qf1</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 1 assays:</span></span>
<span><span class="co">#&gt;  [1] sage: SummarizedExperiment with 977 rows and 11 columns</span></span></code></pre></div>
<p>The piped commands below implement the following worflow:</p>
<ol style="list-style-type: decimal">
<li>Filter decoy features (i.e. those with a label of -1).</li>
<li>Only keep features of rank 1.</li>
<li>Keep features that have a spectrum FDR smaller than 0.05.</li>
<li>Replace 0s by <code>NA</code>, to make missing values explicit.</li>
<li>Log-transform quantitative values.</li>
<li>Normalise data using median centring.</li>
<li>Aggregate features into protein-level quantities by computing the
median and ignoring missing values.</li>
</ol>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf1</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">label</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">|&gt;</span>            <span class="co">## 1</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">rank</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span>            <span class="co">## 2</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">spectrum_fdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="co">## 3</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span>                            <span class="co">## 4</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fl">1</span>, name <span class="op">=</span> <span class="st">"log_sage"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co">## 5</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span>i <span class="op">=</span> <span class="st">"log_sage"</span>,                 <span class="co">## 6</span></span>
<span>              name <span class="op">=</span> <span class="st">"norm_sage"</span>,</span>
<span>              method <span class="op">=</span> <span class="st">"center.median"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span>i <span class="op">=</span> <span class="st">"norm_sage"</span>,        <span class="co">## 7</span></span>
<span>                      name <span class="op">=</span> <span class="st">"proteins"</span>,</span>
<span>                      fcol <span class="op">=</span> <span class="st">"proteins"</span>,</span>
<span>                      fun <span class="op">=</span> <span class="va">colMedians</span>,</span>
<span>                      na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">qf1</span></span>
<span><span class="co">#&gt; 'label' found in 1 out of 1 assay(s)</span></span>
<span><span class="co">#&gt; 'rank' found in 1 out of 1 assay(s)</span></span>
<span><span class="co">#&gt; 'spectrum_fdr' found in 1 out of 1 assay(s)</span></span>
<span><span class="co">#&gt; Your quantitative data contain missing values. Please read the relevant</span></span>
<span><span class="co">#&gt; section(s) in the aggregateFeatures manual page regarding the effects</span></span>
<span><span class="co">#&gt; of missing values on data aggregation.</span></span></code></pre></div>
<p>The new <code>QFeatures</code> object contains 4 assays:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf1</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 4 assays:</span></span>
<span><span class="co">#&gt;  [1] sage: SummarizedExperiment with 309 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [2] log_sage: SummarizedExperiment with 309 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [3] norm_sage: SummarizedExperiment with 309 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [4] proteins: SummarizedExperiment with 297 rows and 11 columns</span></span></code></pre></div>
<p>The pipeline can then be visualised as shown below:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">qf1</span><span class="op">)</span></span></code></pre></div>
<p><img src="sager_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="different-sets-of-samples-per-acquisition">Different sets of samples per acquisition<a class="anchor" aria-label="anchor" href="#different-sets-of-samples-per-acquisition"></a>
</h4>
<p>In this second example, the quantitative data are automatically split
based on the mzML files.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sageQFeatures.html">sageQFeatures</a></span><span class="op">(</span><span class="fu"><a href="../reference/sagerData.html">sagerQuantData</a></span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="../reference/sagerData.html">sagerIdData</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                    byQuant <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"file"</span>, <span class="st">"scannr"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">qf</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 3 assays:</span></span>
<span><span class="co">#&gt;  [1] subset_dq_00084_11cell_90min_hrMS2_A5.mzML: SummarizedExperiment with 288 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [2] subset_dq_00086_11cell_90min_hrMS2_A9.mzML: SummarizedExperiment with 353 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [3] subset_dq_00087_11cell_90min_hrMS2_A11.mzML: SummarizedExperiment with 336 rows and 11 columns</span></span></code></pre></div>
<p>Before processing the data, let’s annotate it. The
<code>renameqf()</code> helper function will automate come of the file
renaming. The annotation will be composed of the file name of origin
(<code>filename</code>), the TMT tag (<code>tmt_tag</code>), and and
acquisition identifier (<code>acquisition</code>).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">renameqf</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"\\.mzML"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"^.+hrMS2_"</span>, <span class="st">""</span>, <span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">qf</span><span class="op">$</span><span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">11</span><span class="op">)</span></span>
<span><span class="va">qf</span><span class="op">$</span><span class="va">tmt_tag</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"tmt_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">11</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">qf</span><span class="op">$</span><span class="va">acquisition</span> <span class="op">&lt;-</span> <span class="fu">renameqf</span><span class="op">(</span><span class="va">qf</span><span class="op">$</span><span class="va">filename</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"psm"</span>, <span class="fu">renameqf</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Below, we update the assay and sample names.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-helpers.html" class="external-link">renamePrimary</a></span><span class="op">(</span><span class="va">qf</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">qf</span><span class="op">$</span><span class="va">acquisition</span>, <span class="va">qf</span><span class="op">$</span><span class="va">tmt_tag</span>, sep <span class="op">=</span> <span class="st">"."</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">CharacterList</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span>, <span class="va">renameqf</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">qf</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 33 rows and 3 columns</span></span>
<span><span class="co">#&gt;                 filename     tmt_tag acquisition</span></span>
<span><span class="co">#&gt;              &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; A5.tmt_1   subset_dq_...       tmt_1          A5</span></span>
<span><span class="co">#&gt; A5.tmt_2   subset_dq_...       tmt_2          A5</span></span>
<span><span class="co">#&gt; A5.tmt_3   subset_dq_...       tmt_3          A5</span></span>
<span><span class="co">#&gt; A5.tmt_4   subset_dq_...       tmt_4          A5</span></span>
<span><span class="co">#&gt; A5.tmt_5   subset_dq_...       tmt_5          A5</span></span>
<span><span class="co">#&gt; ...                  ...         ...         ...</span></span>
<span><span class="co">#&gt; A11.tmt_7  subset_dq_...       tmt_7         A11</span></span>
<span><span class="co">#&gt; A11.tmt_8  subset_dq_...       tmt_8         A11</span></span>
<span><span class="co">#&gt; A11.tmt_9  subset_dq_...       tmt_9         A11</span></span>
<span><span class="co">#&gt; A11.tmt_10 subset_dq_...      tmt_10         A11</span></span>
<span><span class="co">#&gt; A11.tmt_11 subset_dq_...      tmt_11         A11</span></span>
<span><span class="va">qf</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 3 assays:</span></span>
<span><span class="co">#&gt;  [1] psmA5: SummarizedExperiment with 288 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [2] psmA9: SummarizedExperiment with 353 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [3] psmA11: SummarizedExperiment with 336 rows and 11 columns</span></span></code></pre></div>
<p>The piped commands below implement the following worflow:</p>
<ol style="list-style-type: decimal">
<li>Filter decoy features (i.e. those with a label of -1) in all
assays.</li>
<li>Only keep features of rank 1 in all assays.</li>
<li>Keep features that have a spectrum FDR smaller than 0.05 in all
assays.</li>
<li>Replace 0s by <code>NA</code> in all assays, to make missing values
explicit.</li>
<li>Log-transform quantitative values - three new assays are created
from the three initial ones and named with the “log_” prefix.</li>
<li>Aggregate each PSMs into peptide-level quantities by computing the
median and ignoring missing values. The new assays names are defined as
‘peptide’ followed by the orifinal assay identifier.</li>
<li>Join the 3 peptide assays into a new assay that will contain the 33
samples. Missing values are incorporated accordingly, when a peptide is
observed in part of the sets.</li>
<li>Normalise the joined peptide data using median centring.</li>
<li>Aggregate the peptide-level quantities into protein by computing the
median and ignoring missing values. The new assay names will be
‘protein’.</li>
</ol>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">label</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">|&gt;</span>           <span class="co">## 1</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">rank</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span>           <span class="co">## 2</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">spectrum_fdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co">## 3</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span> <span class="op">|&gt;</span>                         <span class="co">## 4</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,                    <span class="co">## 5</span></span>
<span>                 name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"log_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/aggregateFeaturesOverAssays.html" class="external-link">aggregateFeaturesOverAssays</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fl">4</span><span class="op">:</span><span class="fl">6</span>,     <span class="co">## 6</span></span>
<span>                                fcol <span class="op">=</span> <span class="st">"peptide"</span>,</span>
<span>                                name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"psm"</span>, <span class="st">"peptide"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                                fun <span class="op">=</span> <span class="va">colMedians</span>,</span>
<span>                                na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/joinAssays.html" class="external-link">joinAssays</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fl">7</span><span class="op">:</span><span class="fl">9</span>,                      <span class="co">## 7</span></span>
<span>               name <span class="op">=</span> <span class="st">"peptides"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fl">10</span>,                        <span class="co">## 8</span></span>
<span>              name <span class="op">=</span> <span class="st">"norm_peptides"</span>,</span>
<span>              method <span class="op">=</span> <span class="st">"center.median"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span>i <span class="op">=</span> <span class="st">"norm_peptides"</span>,   <span class="co">## 9</span></span>
<span>                      name <span class="op">=</span> <span class="st">"proteins"</span>,</span>
<span>                      fcol <span class="op">=</span> <span class="st">"proteins"</span>,</span>
<span>                      fun <span class="op">=</span> <span class="va">colMedians</span>,</span>
<span>                      na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">qf</span></span>
<span><span class="co">#&gt; 'label' found in 3 out of 3 assay(s)</span></span>
<span><span class="co">#&gt; 'rank' found in 3 out of 3 assay(s)</span></span>
<span><span class="co">#&gt; 'spectrum_fdr' found in 3 out of 3 assay(s)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="va">Aggregated</span><span class="op">:</span> <span class="fl">1</span><span class="op">/</span><span class="fl">3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="va">Aggregated</span><span class="op">:</span> <span class="fl">2</span><span class="op">/</span><span class="fl">3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="va">Aggregated</span><span class="op">:</span> <span class="fl">3</span><span class="op">/</span><span class="fl">3</span></span>
<span><span class="co">#&gt; Your quantitative data contain missing values. Please read the relevant</span></span>
<span><span class="co">#&gt; section(s) in the aggregateFeatures manual page regarding the effects</span></span>
<span><span class="co">#&gt; of missing values on data aggregation.</span></span></code></pre></div>
<p>The final QFeatures object now contains 12 assays:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 12 assays:</span></span>
<span><span class="co">#&gt;  [1] psmA5: SummarizedExperiment with 91 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [2] psmA9: SummarizedExperiment with 109 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [3] psmA11: SummarizedExperiment with 109 rows and 11 columns </span></span>
<span><span class="co">#&gt;  ...</span></span>
<span><span class="co">#&gt;  [10] peptides: SummarizedExperiment with 308 rows and 33 columns </span></span>
<span><span class="co">#&gt;  [11] norm_peptides: SummarizedExperiment with 308 rows and 33 columns </span></span>
<span><span class="co">#&gt;  [12] proteins: SummarizedExperiment with 297 rows and 33 columns</span></span></code></pre></div>
<p>The pipeline can then be visualised as shown below:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span></span></code></pre></div>
<p><img src="sager_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="raw-data">Raw data<a class="anchor" aria-label="anchor" href="#raw-data"></a>
</h3>
<p>We have seen above how to import and process identification and
quantitation data produced by sage using standard R/Bioconductor tools.
Our main goal is to integrate these with the raw MS data that was used
to generate them.</p>
<p>Let’s start by importing these raw data into R as a
<code>Spectra</code> object. The three mzML files can be retrieved with
the <code><a href="../reference/sagerData.html">sagerMzMLData()</a></code> function.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/Spectra" class="external-link">Spectra</a></span><span class="op">)</span></span>
<span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">Spectra</a></span><span class="op">(</span><span class="fu"><a href="../reference/sagerData.html">sagerMzMLData</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sp</span></span>
<span><span class="co">#&gt; MSn data (Spectra) with 598 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">#&gt;       msLevel     rtime scanIndex</span></span>
<span><span class="co">#&gt;     &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt; 1           2   279.184         1</span></span>
<span><span class="co">#&gt; 2           2   731.098         2</span></span>
<span><span class="co">#&gt; 3           2   736.768         3</span></span>
<span><span class="co">#&gt; 4           2   863.610         4</span></span>
<span><span class="co">#&gt; 5           2   866.034         5</span></span>
<span><span class="co">#&gt; ...       ...       ...       ...</span></span>
<span><span class="co">#&gt; 594         2   5155.97       203</span></span>
<span><span class="co">#&gt; 595         2   5163.48       204</span></span>
<span><span class="co">#&gt; 596         2   5178.80       205</span></span>
<span><span class="co">#&gt; 597         2   5179.03       206</span></span>
<span><span class="co">#&gt; 598         2   5249.68       207</span></span>
<span><span class="co">#&gt;  ... 33 more variables/columns.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; file(s):</span></span>
<span><span class="co">#&gt; subset_dq_00084_11cell_90min_hrMS2_A5.mzML</span></span>
<span><span class="co">#&gt; subset_dq_00086_11cell_90min_hrMS2_A9.mzML</span></span>
<span><span class="co">#&gt; subset_dq_00087_11cell_90min_hrMS2_A11.mzML</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="combining-raw-quantitation-and-identification-data">Combining raw, quantitation and identification data<a class="anchor" aria-label="anchor" href="#combining-raw-quantitation-and-identification-data"></a>
</h2>
<div class="section level3">
<h3 id="defining-keys">Defining keys<a class="anchor" aria-label="anchor" href="#defining-keys"></a>
</h3>
<p>We can now create dedicates <em>keys</em> to identify features in the
different data produced above. Let’s start with the spectra within the
<code>sp</code> object. Later, we will create a similar key in the
<code>QFeatures</code> and <code>PSM</code> objects generated above.</p>
<p>The goal of these keys will be to identify matching features across
data types, such as for example to match de MS spectra to the PSMs and
peptides or proteins. These different data types can be stored together
in an <code>MsExperiment</code> object, and matched through one or
multiple keys.</p>
<p>We can add a key (by default, names <code>.KEY</code> that will
identify a feature/scan by concatenating the scan number and the file in
which that scan was acquired.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sp</span><span class="op">$</span><span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">dataOrigin</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keys.html">addKey</a></span><span class="op">(</span><span class="va">sp</span>, vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"filename"</span>, <span class="st">"spectrumId"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sp</span><span class="op">$</span><span class="va">.KEY</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=1004"</span></span>
<span><span class="co">#&gt; [2] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=4181"</span></span>
<span><span class="co">#&gt; [3] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=4224"</span></span>
<span><span class="co">#&gt; [4] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=5124"</span></span>
<span><span class="co">#&gt; [5] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=5144"</span></span>
<span><span class="co">#&gt; [6] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=5517"</span></span></code></pre></div>
<p>Let’s do the same of the PSM object created above:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keys.html">addKey</a></span><span class="op">(</span><span class="va">psm</span>, vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"filename"</span>, <span class="st">"scannr"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">psm</span><span class="op">$</span><span class="va">.KEY</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=31852" </span></span>
<span><span class="co">#&gt; [2] "subset_dq_00087_11cell_90min_hrMS2_A11.mzML.controllerType=0 controllerNumber=1 scan=31961"</span></span>
<span><span class="co">#&gt; [3] "subset_dq_00087_11cell_90min_hrMS2_A11.mzML.controllerType=0 controllerNumber=1 scan=16197"</span></span>
<span><span class="co">#&gt; [4] "subset_dq_00087_11cell_90min_hrMS2_A11.mzML.controllerType=0 controllerNumber=1 scan=32666"</span></span>
<span><span class="co">#&gt; [5] "subset_dq_00086_11cell_90min_hrMS2_A9.mzML.controllerType=0 controllerNumber=1 scan=22318" </span></span>
<span><span class="co">#&gt; [6] "subset_dq_00086_11cell_90min_hrMS2_A9.mzML.controllerType=0 controllerNumber=1 scan=11726"</span></span></code></pre></div>
<p>And finally, with the QFeatures object.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keys.html">addKey</a></span><span class="op">(</span><span class="va">qf</span>, vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"file"</span>, <span class="st">"scannr"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Running <code>addKey</code> on a <code>QFeatures</code> object will
only add a key to assays that do have the relevant variables:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">qf</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="st">".KEY"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;         psmA5         psmA9        psmA11     log_psmA5     log_psmA9 </span></span>
<span><span class="co">#&gt;          TRUE          TRUE          TRUE          TRUE          TRUE </span></span>
<span><span class="co">#&gt;    log_psmA11     peptideA5     peptideA9    peptideA11      peptides </span></span>
<span><span class="co">#&gt;          TRUE          TRUE         FALSE          TRUE         FALSE </span></span>
<span><span class="co">#&gt; norm_peptides      proteins </span></span>
<span><span class="co">#&gt;         FALSE         FALSE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">qf</span><span class="op">[[</span><span class="st">"psmA5"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">.KEY</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=10460"</span></span>
<span><span class="co">#&gt; [2] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=10530"</span></span>
<span><span class="co">#&gt; [3] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=10648"</span></span>
<span><span class="co">#&gt; [4] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=10803"</span></span>
<span><span class="co">#&gt; [5] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=11681"</span></span>
<span><span class="co">#&gt; [6] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=11785"</span></span></code></pre></div>
<p>Let’s now bundle these three types of data together into and
<code>MsExperiment</code> object:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/MsExperiment" class="external-link">MsExperiment</a></span><span class="op">)</span></span>
<span><span class="va">mse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">MsExperiment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sp</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">qdata</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">qf</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">otherData</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span><span class="op">$</span><span class="va">PSM</span> <span class="op">&lt;-</span> <span class="va">psm</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mse</span></span>
<span><span class="co">#&gt; Object of class MsExperiment </span></span>
<span><span class="co">#&gt;  Spectra: MS2 (598) </span></span>
<span><span class="co">#&gt;  QFeatures: 12 assay(s)</span></span>
<span><span class="co">#&gt;  Other data: PSM</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="searching-for-features-across-data-types">Searching for features across data types<a class="anchor" aria-label="anchor" href="#searching-for-features-across-data-types"></a>
</h3>
<p>Each data element <code>mse</code> is referenced by the same key. We
can now query for such keys direcly on the <code>MsExperiment</code>
object, that will percolate the request to its components.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=31852"</span>,</span>
<span>       <span class="st">"subset_dq_00087_11cell_90min_hrMS2_A11.mzML.controllerType=0 controllerNumber=1 scan=32666"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/keys.html">filterKey</a></span><span class="op">(</span><span class="va">mse</span>, <span class="va">k</span><span class="op">)</span></span>
<span><span class="co">#&gt; harmonizing input:</span></span>
<span><span class="co">#&gt;   removing 132 sampleMap rows not in names(experiments)</span></span>
<span><span class="co">#&gt;   removing 11 colData rownames not in sampleMap 'primary'</span></span>
<span><span class="co">#&gt; Object of class MsExperiment </span></span>
<span><span class="co">#&gt;  Spectra: MS2 (2) </span></span>
<span><span class="co">#&gt;  QFeatures: 6 assay(s)</span></span>
<span><span class="co">#&gt;  Other data: PSM</span></span></code></pre></div>
<p>It is however not convenient to search an arbitrary key. The more
useful usecase would be to look for a protein of interest, for example,
and then identify the features across other data types using the
relevant key(s).</p>
<p>Let’s imagine that protein <em>DNLI3_HUMAN</em> is of interest to us.
It is referenced as <code>"sp|P49916|DNLI3_HUMAN"</code> in our data.
The <code>dropEmptyAssay()</code> simply drops assays that didn’t
reference the protein of interest - it is used to simplify the output
here.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">qdata</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span><span class="op">[</span><span class="st">"sp|P49916|DNLI3_HUMAN"</span>, , <span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">dropEmptyAssays</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: 'experiments' dropped; see 'drops()'</span></span>
<span><span class="co">#&gt; harmonizing input:</span></span>
<span><span class="co">#&gt;   removing 66 sampleMap rows not in names(experiments)</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 6 assays:</span></span>
<span><span class="co">#&gt;  [1] psmA5: SummarizedExperiment with 1 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [2] log_psmA5: SummarizedExperiment with 1 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [3] peptideA5: SummarizedExperiment with 1 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [4] peptides: SummarizedExperiment with 1 rows and 33 columns </span></span>
<span><span class="co">#&gt;  [5] norm_peptides: SummarizedExperiment with 1 rows and 33 columns </span></span>
<span><span class="co">#&gt;  [6] proteins: SummarizedExperiment with 1 rows and 33 columns</span></span></code></pre></div>
<p>These are features that we might be interested in tracking across all
data using their key… which we can easily retrieve with the
<code><a href="../reference/keys.html">getKey()</a></code> function:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">qdata</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span><span class="op">[</span><span class="st">"sp|P49916|DNLI3_HUMAN"</span>, , <span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">dropEmptyAssays</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/keys.html">getKey</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: 'experiments' dropped; see 'drops()'</span></span>
<span><span class="co">#&gt; List of length 3</span></span>
<span><span class="co">#&gt; names(3): psmA5 log_psmA5 peptideA5</span></span></code></pre></div>
<p>Given that <code><a href="../reference/keys.html">getKey()</a></code> is run on a <code>QFeatures</code>
object, it returns the keys of all assays. Let’s extract the unique keys
as a character:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_key</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">qdata</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span><span class="op">[</span><span class="st">"sp|P49916|DNLI3_HUMAN"</span>, , <span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">dropEmptyAssays</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/keys.html">getKey</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: 'experiments' dropped; see 'drops()'</span></span>
<span><span class="va">my_key</span></span>
<span><span class="co">#&gt; [1] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=17928"</span></span></code></pre></div>
<p>And now use <code>mykey</code> to subset the different component of
the <code>MsExperiment</code> object:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_mse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keys.html">filterKey</a></span><span class="op">(</span><span class="va">mse</span>, <span class="va">my_key</span><span class="op">)</span></span>
<span><span class="co">#&gt; harmonizing input:</span></span>
<span><span class="co">#&gt;   removing 165 sampleMap rows not in names(experiments)</span></span>
<span><span class="co">#&gt;   removing 22 colData rownames not in sampleMap 'primary'</span></span>
<span><span class="va">my_mse</span></span>
<span><span class="co">#&gt; Object of class MsExperiment </span></span>
<span><span class="co">#&gt;  Spectra: MS2 (1) </span></span>
<span><span class="co">#&gt;  QFeatures: 3 assay(s)</span></span>
<span><span class="co">#&gt;  Other data: PSM</span></span></code></pre></div>
<p>From the above, we can see that <em>DNLI3_HUMAN</em> was identified
and quantified by a single peptides, that has the following quantitation
profile acorss the 11 samples…</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assay</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">qdata</a></span><span class="op">(</span><span class="va">my_mse</span><span class="op">)</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                  A5.tmt_1 A5.tmt_2 A5.tmt_3 A5.tmt_4 A5.tmt_5</span></span>
<span><span class="co">#&gt; [+229.1629]-SATNLPQLK[+229.1629] 18.48759 19.64292      NaN      NaN  18.4555</span></span>
<span><span class="co">#&gt;                                  A5.tmt_6 A5.tmt_7 A5.tmt_8 A5.tmt_9 A5.tmt_10</span></span>
<span><span class="co">#&gt; [+229.1629]-SATNLPQLK[+229.1629]      NaN 20.28052 18.19301      NaN  18.82892</span></span>
<span><span class="co">#&gt;                                  A5.tmt_11</span></span>
<span><span class="co">#&gt; [+229.1629]-SATNLPQLK[+229.1629]       NaN</span></span></code></pre></div>
<p>… and that that peptides was matched by a single spectrum, shown
below:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectra</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">my_mse</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="sager_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="misc">Misc<a class="anchor" aria-label="anchor" href="#misc"></a>
</h2>
<div class="section level3">
<h3 id="getting-help">Getting help<a class="anchor" aria-label="anchor" href="#getting-help"></a>
</h3>
<p>Please <a href="https://github.com/UCLouvain-CBIO/sager/issues" class="external-link">open
an issue</a> on the package’s Github repository.</p>
<!-- ## Citing `sager` -->
<!-- ```{r citation} -->
<!-- citation("sager") -->
<!-- ``` -->
</div>
<div class="section level3">
<h3 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h3>
<pre><code><span><span class="co">#&gt; R Under development (unstable) (2023-11-02 r85465)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] MsExperiment_1.5.1          Spectra_1.13.0             </span></span>
<span><span class="co">#&gt;  [3] ProtGenerics_1.35.0         BiocParallel_1.37.0        </span></span>
<span><span class="co">#&gt;  [5] scp_1.13.0                  QFeatures_1.11.3           </span></span>
<span><span class="co">#&gt;  [7] MultiAssayExperiment_1.29.0 SummarizedExperiment_1.33.0</span></span>
<span><span class="co">#&gt;  [9] Biobase_2.63.0              GenomicRanges_1.55.1       </span></span>
<span><span class="co">#&gt; [11] GenomeInfoDb_1.39.0         IRanges_2.37.0             </span></span>
<span><span class="co">#&gt; [13] MatrixGenerics_1.15.0       matrixStats_1.1.0          </span></span>
<span><span class="co">#&gt; [15] ggplot2_3.4.4               PSMatch_1.7.0              </span></span>
<span><span class="co">#&gt; [17] S4Vectors_0.41.1            BiocGenerics_0.49.1        </span></span>
<span><span class="co">#&gt; [19] sager_0.3.1                 BiocStyle_2.31.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] DBI_1.1.3                   bitops_1.0-7               </span></span>
<span><span class="co">#&gt;  [3] rlang_1.1.2                 magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;  [5] clue_0.3-65                 compiler_4.4.0             </span></span>
<span><span class="co">#&gt;  [7] RSQLite_2.3.3               systemfonts_1.0.5          </span></span>
<span><span class="co">#&gt;  [9] vctrs_0.6.4                 stringr_1.5.0              </span></span>
<span><span class="co">#&gt; [11] pkgconfig_2.0.3             MetaboCoreUtils_1.11.0     </span></span>
<span><span class="co">#&gt; [13] crayon_1.5.2                fastmap_1.1.1              </span></span>
<span><span class="co">#&gt; [15] dbplyr_2.4.0                XVector_0.43.0             </span></span>
<span><span class="co">#&gt; [17] labeling_0.4.3              utf8_1.2.4                 </span></span>
<span><span class="co">#&gt; [19] rmarkdown_2.25              ragg_1.2.6                 </span></span>
<span><span class="co">#&gt; [21] purrr_1.0.2                 bit_4.0.5                  </span></span>
<span><span class="co">#&gt; [23] xfun_0.41                   zlibbioc_1.49.0            </span></span>
<span><span class="co">#&gt; [25] cachem_1.0.8                jsonlite_1.8.7             </span></span>
<span><span class="co">#&gt; [27] blob_1.2.4                  highr_0.10                 </span></span>
<span><span class="co">#&gt; [29] DelayedArray_0.29.0         parallel_4.4.0             </span></span>
<span><span class="co">#&gt; [31] cluster_2.1.4               R6_2.5.1                   </span></span>
<span><span class="co">#&gt; [33] bslib_0.5.1                 stringi_1.7.12             </span></span>
<span><span class="co">#&gt; [35] jquerylib_0.1.4             Rcpp_1.0.11                </span></span>
<span><span class="co">#&gt; [37] bookdown_0.36               knitr_1.45                 </span></span>
<span><span class="co">#&gt; [39] BiocBaseUtils_1.5.0         Matrix_1.6-2               </span></span>
<span><span class="co">#&gt; [41] igraph_1.5.1                tidyselect_1.2.0           </span></span>
<span><span class="co">#&gt; [43] abind_1.4-5                 yaml_2.3.7                 </span></span>
<span><span class="co">#&gt; [45] codetools_0.2-19            curl_5.1.0                 </span></span>
<span><span class="co">#&gt; [47] lattice_0.22-5              tibble_3.2.1               </span></span>
<span><span class="co">#&gt; [49] withr_2.5.2                 evaluate_0.23              </span></span>
<span><span class="co">#&gt; [51] desc_1.4.2                  BiocFileCache_2.11.1       </span></span>
<span><span class="co">#&gt; [53] pillar_1.9.0                BiocManager_1.30.22        </span></span>
<span><span class="co">#&gt; [55] filelock_1.0.2              ncdf4_1.21                 </span></span>
<span><span class="co">#&gt; [57] generics_0.1.3              rprojroot_2.0.4            </span></span>
<span><span class="co">#&gt; [59] RCurl_1.98-1.13             munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [61] scales_1.2.1                glue_1.6.2                 </span></span>
<span><span class="co">#&gt; [63] lazyeval_0.2.2              tools_4.4.0                </span></span>
<span><span class="co">#&gt; [65] mzR_2.37.0                  fs_1.6.3                   </span></span>
<span><span class="co">#&gt; [67] grid_4.4.0                  MsCoreUtils_1.15.1         </span></span>
<span><span class="co">#&gt; [69] colorspace_2.1-0            SingleCellExperiment_1.25.0</span></span>
<span><span class="co">#&gt; [71] GenomeInfoDbData_1.2.11     cli_3.6.1                  </span></span>
<span><span class="co">#&gt; [73] textshaping_0.3.7           fansi_1.0.5                </span></span>
<span><span class="co">#&gt; [75] S4Arrays_1.3.0              dplyr_1.1.3                </span></span>
<span><span class="co">#&gt; [77] AnnotationFilter_1.27.0     gtable_0.3.4               </span></span>
<span><span class="co">#&gt; [79] sass_0.4.7                  digest_0.6.33              </span></span>
<span><span class="co">#&gt; [81] SparseArray_1.3.1           farver_2.1.1               </span></span>
<span><span class="co">#&gt; [83] memoise_2.0.1               htmltools_0.5.7            </span></span>
<span><span class="co">#&gt; [85] pkgdown_2.0.7.9000          lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt; [87] httr_1.4.7                  bit64_4.0.5                </span></span>
<span><span class="co">#&gt; [89] MASS_7.3-60.1</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
