<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using the `sager` package to import and analyse `sage` results in R • sager</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using the `sager` package to import and analyse `sage` results in R">
<meta property="og:description" content="sager">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sager</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/sager.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/UCLouvain-CBIO/sager/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using the <code>sager</code> package to import
and analyse <code>sage</code> results in R</h1>
                        <h4 data-toc-skip class="author">Laurent
Gatto</h4>
            <address class="author_afil">
      Computational Biology and Bioinformatics,
UCLouvain<br><a class="author_email" href="mailto:#"></a><a href="mailto:laurent.gatto@uclouvain.be" class="email">laurent.gatto@uclouvain.be</a>
      </address>
                  
            <h4 data-toc-skip class="date">17 April 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/UCLouvain-CBIO/sager/blob/HEAD/vignettes/sager.Rmd" class="external-link"><code>vignettes/sager.Rmd</code></a></small>
      <div class="hidden name"><code>sager.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="section level3">
<h3 id="the-sager-package">The <code>sager</code> package<a class="anchor" aria-label="anchor" href="#the-sager-package"></a>
</h3>
<p><a href="https://lazear.github.io/sage/" class="external-link">Sage</a> is a new
cross-platform, extremely performant, open source proteomics search
engine <a href="https://github.com/lazear" class="external-link">Michael Lazear</a>. It uses
mzML files as input, is parameterised by a <a href="https://github.com/lazear/sage/blob/master/DOCS.md" class="external-link">json file</a>
and is executed from the command line. In addition to peptide
identification, sage also includes a variety of advanced features such
as retention time prediction and quantification (both isobaric &amp;
LFQ). It produces two files: <code>results.sage.tsv</code> with the
identification results and, when configured to quantify features,
<code>quant.tsv</code>.</p>
<p>The <code>sager</code> package uses these two files as input. It
facilitates their import into established Bioconductor classes:</p>
<ul>
<li><p>Identification results are parsed and imported as <a href="https://rformassspectrometry.github.io/PSMatch/" class="external-link">PSMatch::PSM()</a>
objects with <code><a href="../reference/sagePSM.html">sagePSM()</a></code>.</p></li>
<li><p>Quantitation (and identification) results are parsed, merged and
imported as <a href="https://rformassspectrometry.github.io/QFeatures/articles/QFeatures.html" class="external-link">QFeatures::QFeatures()</a>
objects with <code><a href="../reference/sageQFeatures.html">sageQFeatures()</a></code>.</p></li>
<li><p>Sage results can also be integrated with a <a href="https://rformassspectrometry.github.io/Spectra/" class="external-link">Spectra::Spectra()</a>
instance (holding the raw data) into a <a href="https://rformassspectrometry.github.io/MsExperiment/" class="external-link">MsExperiment::MsExperiment()</a>
object, as documented in this document.</p></li>
</ul>
<p>The goal of the <code>sager</code> package can thus be summarised as
a way to leverage the existing <a href="https://www.rformassspectrometry.org/" class="external-link">R for Mass Spectrometry</a>
and <a href="https://bioconductor.org/" class="external-link">Bioconductor</a> infrastructure
to analyse sage results.</p>
</div>
<div class="section level3">
<h3 id="installation-instructions">Installation instructions<a class="anchor" aria-label="anchor" href="#installation-instructions"></a>
</h3>
<p>To install <code>sager</code> and its dependencies, you’ll need the
<code>BiocManager</code> package, that can be installed from CRAN
(unless you already have it):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre></div>
<p>Note that <code>sager</code> is under constant development, and thus
will depend on some under-development dependencies, that haven’t made it
into the stable releases yet. Here are the instructions to install
these:</p>
<p>To install the <em>devel</em> version of <code>PSMatch</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"RforMassSpectrometry/PSMatch"</span><span class="op">)</span></span></code></pre></div>
<p>To install <code>sager</code> and its dependencies, you can run the
following command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"UCLouvain-CBIO/sager"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="analysing-sage-results">Analysing <code>sage</code> results<a class="anchor" aria-label="anchor" href="#analysing-sage-results"></a>
</h2>
<div class="section level3">
<h3 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">sager</span><span class="op">)</span></span></code></pre></div>
<p>The <code>sager</code> package provides example data</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sagerData.html">sagerAvailableData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; quant    id  mzml </span></span>
<span><span class="co">#&gt;  TRUE  TRUE  TRUE</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sagerData.html">sagerAddData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'quant' already available.</span></span>
<span><span class="co">#&gt; 'id' already available.</span></span>
<span><span class="co">#&gt; 'mzml' already available.</span></span></code></pre></div>
<p>For details and provenance of the data, please see the
<code><a href="../reference/sagerData.html">?sagerData</a></code> manual page.</p>
</div>
<div class="section level3">
<h3 id="identification-data">Identification data<a class="anchor" aria-label="anchor" href="#identification-data"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/PSM" class="external-link">PSMatch</a></span><span class="op">)</span></span>
<span><span class="va">psm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sagePSM.html">sagePSM</a></span><span class="op">(</span><span class="fu"><a href="../reference/sagerData.html">sagerIdData</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">psm</span></span>
<span><span class="co">#&gt; PSM with 977 rows and 34 columns.</span></span>
<span><span class="co">#&gt; Spectra: 595 unique</span></span>
<span><span class="co">#&gt;   db: 679 target, 298 decoy</span></span>
<span><span class="co">#&gt;   ranks: 1:977 </span></span>
<span><span class="co">#&gt; Peptides: 976 unique, 0 multiple</span></span>
<span><span class="co">#&gt; Proteins: 947 </span></span>
<span><span class="co">#&gt; names(34): peptide proteins ... protein_fdr ms1_intensity</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">psm</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">hyperscore</span>,</span>
<span>               colour <span class="op">=</span> <span class="va">label</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="sager_files/figure-html/ggplot-1.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">psm</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">spectrum_fdr</span>,</span>
<span>               colour <span class="op">=</span> <span class="va">label</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="sager_files/figure-html/ggplog2-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PSMatch/man/filterPSMs.html" class="external-link">filterPSMs</a></span><span class="op">(</span><span class="va">psm</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/PSMatch/man/filterPSMs.html" class="external-link">filterPsmFdr</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Starting with 977 PSMs:</span></span>
<span><span class="co">#&gt; Removed 298 decoy hits.</span></span>
<span><span class="co">#&gt; Removed 0 PSMs with rank &gt; 1.</span></span>
<span><span class="co">#&gt; Removed 0 shared peptides.</span></span>
<span><span class="co">#&gt; 679 PSMs left.</span></span>
<span><span class="co">#&gt; Removed 370 PSMs with FDR &lt; spectrum_fdr.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="quantitative-data">Quantitative data<a class="anchor" aria-label="anchor" href="#quantitative-data"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/QFeatures" class="external-link">QFeatures</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://UCLouvain-CBIO.github.io/scp" class="external-link">scp</a></span><span class="op">)</span></span></code></pre></div>
<p>Multiple fractions for the same set of samples:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sageQFeatures.html">sageQFeatures</a></span><span class="op">(</span><span class="fu"><a href="../reference/sagerData.html">sagerQuantData</a></span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="../reference/sagerData.html">sagerIdData</a></span><span class="op">(</span><span class="op">)</span>, splitBy <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">qf1</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 1 assays:</span></span>
<span><span class="co">#&gt;  [1] sage: SummarizedExperiment with 977 rows and 11 columns</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf1</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">label</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">rank</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">spectrum_fdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fl">1</span>, name <span class="op">=</span> <span class="st">"log_sage"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span>i <span class="op">=</span> <span class="st">"log_sage"</span>,</span>
<span>              name <span class="op">=</span> <span class="st">"norm_sage"</span>,</span>
<span>              method <span class="op">=</span> <span class="st">"center.mean"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span>i <span class="op">=</span> <span class="st">"norm_sage"</span>,</span>
<span>                      name <span class="op">=</span> <span class="st">"proteins"</span>,</span>
<span>                      fcol <span class="op">=</span> <span class="st">"proteins"</span>,</span>
<span>                      fun <span class="op">=</span> <span class="va">colMedians</span>,</span>
<span>                      na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">qf1</span></span>
<span><span class="co">#&gt; 'label' found in 1 out of 1 assay(s)</span></span>
<span><span class="co">#&gt; 'rank' found in 1 out of 1 assay(s)</span></span>
<span><span class="co">#&gt; 'spectrum_fdr' found in 1 out of 1 assay(s)</span></span>
<span><span class="co">#&gt; Your quantitative data contain missing values. Please read the relevant</span></span>
<span><span class="co">#&gt; section(s) in the aggregateFeatures manual page regarding the effects</span></span>
<span><span class="co">#&gt; of missing values on data aggregation.</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">qf1</span><span class="op">)</span></span></code></pre></div>
<p><img src="sager_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf1</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 4 assays:</span></span>
<span><span class="co">#&gt;  [1] sage: SummarizedExperiment with 309 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [2] log_sage: SummarizedExperiment with 309 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [3] norm_sage: SummarizedExperiment with 309 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [4] proteins: SummarizedExperiment with 297 rows and 11 columns</span></span></code></pre></div>
<p>Different sets of samples per acquisition:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sageQFeatures.html">sageQFeatures</a></span><span class="op">(</span><span class="fu"><a href="../reference/sagerData.html">sagerQuantData</a></span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="../reference/sagerData.html">sagerIdData</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">qf</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 3 assays:</span></span>
<span><span class="co">#&gt;  [1] subset_dq_00084_11cell_90min_hrMS2_A5.mzML: SummarizedExperiment with 288 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [2] subset_dq_00086_11cell_90min_hrMS2_A9.mzML: SummarizedExperiment with 353 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [3] subset_dq_00087_11cell_90min_hrMS2_A11.mzML: SummarizedExperiment with 336 rows and 11 columns</span></span></code></pre></div>
<p>Annotate data:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">renameqf</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"\\.mzML"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"^.+hrMS2_"</span>, <span class="st">""</span>, <span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">qf</span><span class="op">$</span><span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">11</span><span class="op">)</span></span>
<span><span class="va">qf</span><span class="op">$</span><span class="va">tmt_tag</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"tmt_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">11</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">qf</span><span class="op">$</span><span class="va">acquisition</span> <span class="op">&lt;-</span> <span class="fu">renameqf</span><span class="op">(</span><span class="va">qf</span><span class="op">$</span><span class="va">filename</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"psm"</span>, <span class="fu">renameqf</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">qf</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-helpers.html" class="external-link">renamePrimary</a></span><span class="op">(</span><span class="va">qf</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">qf</span><span class="op">$</span><span class="va">acquisition</span>, <span class="va">qf</span><span class="op">$</span><span class="va">tmt_tag</span>, sep <span class="op">=</span> <span class="st">"."</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">CharacterList</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span>, <span class="va">renameqf</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 3 assays:</span></span>
<span><span class="co">#&gt;  [1] psmA5: SummarizedExperiment with 288 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [2] psmA9: SummarizedExperiment with 353 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [3] psmA11: SummarizedExperiment with 336 rows and 11 columns</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">qf</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 33 rows and 3 columns</span></span>
<span><span class="co">#&gt;                 filename     tmt_tag acquisition</span></span>
<span><span class="co">#&gt;              &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; A5.tmt_1   subset_dq_...       tmt_1          A5</span></span>
<span><span class="co">#&gt; A5.tmt_2   subset_dq_...       tmt_2          A5</span></span>
<span><span class="co">#&gt; A5.tmt_3   subset_dq_...       tmt_3          A5</span></span>
<span><span class="co">#&gt; A5.tmt_4   subset_dq_...       tmt_4          A5</span></span>
<span><span class="co">#&gt; A5.tmt_5   subset_dq_...       tmt_5          A5</span></span>
<span><span class="co">#&gt; ...                  ...         ...         ...</span></span>
<span><span class="co">#&gt; A11.tmt_7  subset_dq_...       tmt_7         A11</span></span>
<span><span class="co">#&gt; A11.tmt_8  subset_dq_...       tmt_8         A11</span></span>
<span><span class="co">#&gt; A11.tmt_9  subset_dq_...       tmt_9         A11</span></span>
<span><span class="co">#&gt; A11.tmt_10 subset_dq_...      tmt_10         A11</span></span>
<span><span class="co">#&gt; A11.tmt_11 subset_dq_...      tmt_11         A11</span></span></code></pre></div>
<p>Data processing:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">label</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">rank</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">spectrum_fdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"log_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/aggregateFeaturesOverAssays.html" class="external-link">aggregateFeaturesOverAssays</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fl">4</span><span class="op">:</span><span class="fl">6</span>,</span>
<span>                                fcol <span class="op">=</span> <span class="st">"peptide"</span>,</span>
<span>                                name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"psm"</span>, <span class="st">"peptide"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                                fun <span class="op">=</span> <span class="va">colMedians</span>,</span>
<span>                                na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/joinAssays.html" class="external-link">joinAssays</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fl">7</span><span class="op">:</span><span class="fl">9</span>,</span>
<span>               name <span class="op">=</span> <span class="st">"peptides"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fl">10</span>,</span>
<span>              name <span class="op">=</span> <span class="st">"norm_peptides"</span>,</span>
<span>              method <span class="op">=</span> <span class="st">"center.mean"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span>i <span class="op">=</span> <span class="st">"norm_peptides"</span>,</span>
<span>                      name <span class="op">=</span> <span class="st">"proteins"</span>,</span>
<span>                      fcol <span class="op">=</span> <span class="st">"proteins"</span>,</span>
<span>                      fun <span class="op">=</span> <span class="va">colMedians</span>,</span>
<span>                      na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">qf</span></span>
<span><span class="co">#&gt; 'label' found in 3 out of 3 assay(s)</span></span>
<span><span class="co">#&gt; 'rank' found in 3 out of 3 assay(s)</span></span>
<span><span class="co">#&gt; 'spectrum_fdr' found in 3 out of 3 assay(s)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="va">Aggregated</span><span class="op">:</span> <span class="fl">1</span><span class="op">/</span><span class="fl">3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="va">Aggregated</span><span class="op">:</span> <span class="fl">2</span><span class="op">/</span><span class="fl">3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="va">Aggregated</span><span class="op">:</span> <span class="fl">3</span><span class="op">/</span><span class="fl">3</span></span>
<span><span class="co">#&gt; Your quantitative data contain missing values. Please read the relevant</span></span>
<span><span class="co">#&gt; section(s) in the aggregateFeatures manual page regarding the effects</span></span>
<span><span class="co">#&gt; of missing values on data aggregation.</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">qf</span><span class="op">)</span></span></code></pre></div>
<p><img src="sager_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 12 assays:</span></span>
<span><span class="co">#&gt;  [1] psmA5: SummarizedExperiment with 91 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [2] psmA9: SummarizedExperiment with 109 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [3] psmA11: SummarizedExperiment with 109 rows and 11 columns </span></span>
<span><span class="co">#&gt;  ...</span></span>
<span><span class="co">#&gt;  [10] peptides: SummarizedExperiment with 308 rows and 33 columns </span></span>
<span><span class="co">#&gt;  [11] norm_peptides: SummarizedExperiment with 308 rows and 33 columns </span></span>
<span><span class="co">#&gt;  [12] proteins: SummarizedExperiment with 297 rows and 33 columns</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="raw-data">Raw data<a class="anchor" aria-label="anchor" href="#raw-data"></a>
</h3>
<p>We have seen above how to import and process identification and
quantitation data produced by sage using standard R/Bioconductor tools.
Our main goal is to integrate these with the raw MS data that was used
to generate them.</p>
<p>Let’s start by importing these raw data into R as a
<code>Spectra</code> object. The three mzML files can be retrieved with
the <code><a href="../reference/sagerData.html">sagerMzMLData()</a></code> function.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/Spectra" class="external-link">Spectra</a></span><span class="op">)</span></span>
<span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">Spectra</a></span><span class="op">(</span><span class="fu"><a href="../reference/sagerData.html">sagerMzMLData</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sp</span></span>
<span><span class="co">#&gt; MSn data (Spectra) with 598 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">#&gt;       msLevel     rtime scanIndex</span></span>
<span><span class="co">#&gt;     &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt; 1           2   279.184         1</span></span>
<span><span class="co">#&gt; 2           2   731.098         2</span></span>
<span><span class="co">#&gt; 3           2   736.768         3</span></span>
<span><span class="co">#&gt; 4           2   863.610         4</span></span>
<span><span class="co">#&gt; 5           2   866.034         5</span></span>
<span><span class="co">#&gt; ...       ...       ...       ...</span></span>
<span><span class="co">#&gt; 594         2   5155.97       203</span></span>
<span><span class="co">#&gt; 595         2   5163.48       204</span></span>
<span><span class="co">#&gt; 596         2   5178.80       205</span></span>
<span><span class="co">#&gt; 597         2   5179.03       206</span></span>
<span><span class="co">#&gt; 598         2   5249.68       207</span></span>
<span><span class="co">#&gt;  ... 33 more variables/columns.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; file(s):</span></span>
<span><span class="co">#&gt; subset_dq_00084_11cell_90min_hrMS2_A5.mzML</span></span>
<span><span class="co">#&gt; subset_dq_00086_11cell_90min_hrMS2_A9.mzML</span></span>
<span><span class="co">#&gt; subset_dq_00087_11cell_90min_hrMS2_A11.mzML</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="combining-raw-quantitation-and-identification-data">Combining raw, quantitation and identification data<a class="anchor" aria-label="anchor" href="#combining-raw-quantitation-and-identification-data"></a>
</h2>
<div class="section level3">
<h3 id="defining-keys">Defining keys<a class="anchor" aria-label="anchor" href="#defining-keys"></a>
</h3>
<p>We can now create dedicates <em>keys</em> to identify features in the
different data produced above. Let’s start with the spectra within the
<code>sp</code> object. Later, we will create a similar key in the
<code>QFeatures</code> and <code>PSM</code> objects generated above.</p>
<p>The goal of these keys will be to identify matching features across
data types, such as for example to match de MS spectra to the PSMs and
peptides or proteins. These different data types can be stored together
in an <code>MsExperiment</code> object, and matched through one or
multiple keys.</p>
<p>We can add a key (by default, names <code>.KEY</code> that will
identify a feature/scan by concatenating the scan number and the file in
which that scan was acquired.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sp</span><span class="op">$</span><span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">dataOrigin</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keys.html">addKEY</a></span><span class="op">(</span><span class="va">sp</span>, vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"filename"</span>, <span class="st">"spectrumId"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sp</span><span class="op">$</span><span class="va">.KEY</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=1004"</span></span>
<span><span class="co">#&gt; [2] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=4181"</span></span>
<span><span class="co">#&gt; [3] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=4224"</span></span>
<span><span class="co">#&gt; [4] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=5124"</span></span>
<span><span class="co">#&gt; [5] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=5144"</span></span>
<span><span class="co">#&gt; [6] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=5517"</span></span></code></pre></div>
<p>Let’s do the same of the PSM object created above:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keys.html">addKEY</a></span><span class="op">(</span><span class="va">psm</span>, vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"filename"</span>, <span class="st">"scannr"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">psm</span><span class="op">$</span><span class="va">.KEY</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=31852" </span></span>
<span><span class="co">#&gt; [2] "subset_dq_00087_11cell_90min_hrMS2_A11.mzML.controllerType=0 controllerNumber=1 scan=31961"</span></span>
<span><span class="co">#&gt; [3] "subset_dq_00087_11cell_90min_hrMS2_A11.mzML.controllerType=0 controllerNumber=1 scan=16197"</span></span>
<span><span class="co">#&gt; [4] "subset_dq_00087_11cell_90min_hrMS2_A11.mzML.controllerType=0 controllerNumber=1 scan=32666"</span></span>
<span><span class="co">#&gt; [5] "subset_dq_00086_11cell_90min_hrMS2_A9.mzML.controllerType=0 controllerNumber=1 scan=22318" </span></span>
<span><span class="co">#&gt; [6] "subset_dq_00086_11cell_90min_hrMS2_A9.mzML.controllerType=0 controllerNumber=1 scan=11726"</span></span></code></pre></div>
<p>And finally, with the QFeatures object.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keys.html">addKEY</a></span><span class="op">(</span><span class="va">qf</span>, vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"file"</span>, <span class="st">"scannr"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Running <code>addKEY</code> on a <code>QFeatures</code> object will
only add a key to assays that do have the relevant variables:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">qf</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="st">".KEY"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;         psmA5         psmA9        psmA11     log_psmA5     log_psmA9 </span></span>
<span><span class="co">#&gt;          TRUE          TRUE          TRUE          TRUE          TRUE </span></span>
<span><span class="co">#&gt;    log_psmA11     peptideA5     peptideA9    peptideA11      peptides </span></span>
<span><span class="co">#&gt;          TRUE          TRUE         FALSE          TRUE         FALSE </span></span>
<span><span class="co">#&gt; norm_peptides      proteins </span></span>
<span><span class="co">#&gt;         FALSE         FALSE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">qf</span><span class="op">[[</span><span class="st">"psmA5"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">.KEY</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                                                           4 </span></span>
<span><span class="co">#&gt; "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=10460" </span></span>
<span><span class="co">#&gt;                                                                                           5 </span></span>
<span><span class="co">#&gt; "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=10530" </span></span>
<span><span class="co">#&gt;                                                                                           8 </span></span>
<span><span class="co">#&gt; "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=10648" </span></span>
<span><span class="co">#&gt;                                                                                          10 </span></span>
<span><span class="co">#&gt; "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=10803" </span></span>
<span><span class="co">#&gt;                                                                                          17 </span></span>
<span><span class="co">#&gt; "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=11681" </span></span>
<span><span class="co">#&gt;                                                                                          19 </span></span>
<span><span class="co">#&gt; "subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=11785"</span></span></code></pre></div>
<p>Let’s now bundle these three types of data together into and
<code>MsExperiment</code> object:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/MsExperiment" class="external-link">MsExperiment</a></span><span class="op">)</span></span>
<span><span class="va">mse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">MsExperiment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sp</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">qdata</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">qf</span></span>
<span><span class="va">mse</span><span class="op">@</span><span class="va">otherData</span><span class="op">$</span><span class="va">PSM</span> <span class="op">&lt;-</span> <span class="va">psm</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mse</span></span>
<span><span class="co">#&gt; Object of class MsExperiment </span></span>
<span><span class="co">#&gt;  Files:  </span></span>
<span><span class="co">#&gt;  Spectra: MS2 (598)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="searching-for-keys-across-data-types">Searching for keys across data types<a class="anchor" aria-label="anchor" href="#searching-for-keys-across-data-types"></a>
</h3>
<p>Each data element <code>mse</code> is referenced by the same key. We
can now query for such keys direcly on the <code>MsExperiment</code>
object, that will percolate the request to its components.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"subset_dq_00084_11cell_90min_hrMS2_A5.mzML.controllerType=0 controllerNumber=1 scan=31852"</span>,</span>
<span>       <span class="st">"subset_dq_00087_11cell_90min_hrMS2_A11.mzML.controllerType=0 controllerNumber=1 scan=32666"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/keys.html">subsetByKEY</a></span><span class="op">(</span><span class="va">mse</span>, <span class="va">k</span><span class="op">)</span></span>
<span><span class="co">#&gt; harmonizing input:</span></span>
<span><span class="co">#&gt;   removing 132 sampleMap rows not in names(experiments)</span></span>
<span><span class="co">#&gt;   removing 11 colData rownames not in sampleMap 'primary'</span></span>
<span><span class="co">#&gt; Object of class MsExperiment </span></span>
<span><span class="co">#&gt;  Files:  </span></span>
<span><span class="co">#&gt;  Spectra: MS2 (2)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="misc">Misc<a class="anchor" aria-label="anchor" href="#misc"></a>
</h2>
<div class="section level3">
<h3 id="getting-help">Getting help<a class="anchor" aria-label="anchor" href="#getting-help"></a>
</h3>
<p>Please <a href="https://github.com/UCLouvain-CBIO/sager/issues" class="external-link">open
an issue</a> on the package’s Github repository.</p>
<!-- ## Citing `sager` -->
<!-- ```{r citation} -->
<!-- citation("sager") -->
<!-- ``` -->
</div>
<div class="section level3">
<h3 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h3>
<pre><code><span><span class="co">#&gt; R version 4.3.0 beta (2023-04-12 r84240)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] MsExperiment_1.1.3          Spectra_1.9.15             </span></span>
<span><span class="co">#&gt;  [3] ProtGenerics_1.31.0         BiocParallel_1.33.12       </span></span>
<span><span class="co">#&gt;  [5] scp_1.9.1                   QFeatures_1.9.3            </span></span>
<span><span class="co">#&gt;  [7] MultiAssayExperiment_1.25.9 SummarizedExperiment_1.29.1</span></span>
<span><span class="co">#&gt;  [9] Biobase_2.59.0              GenomicRanges_1.51.4       </span></span>
<span><span class="co">#&gt; [11] GenomeInfoDb_1.35.17        IRanges_2.33.1             </span></span>
<span><span class="co">#&gt; [13] MatrixGenerics_1.11.1       matrixStats_0.63.0         </span></span>
<span><span class="co">#&gt; [15] ggplot2_3.4.2               PSMatch_1.3.3              </span></span>
<span><span class="co">#&gt; [17] S4Vectors_0.37.6            BiocGenerics_0.45.3        </span></span>
<span><span class="co">#&gt; [19] sager_0.2.1                 BiocStyle_2.27.2           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.0            farver_2.1.1               </span></span>
<span><span class="co">#&gt;  [3] dplyr_1.1.1                 blob_1.2.4                 </span></span>
<span><span class="co">#&gt;  [5] filelock_1.0.2              bitops_1.0-7               </span></span>
<span><span class="co">#&gt;  [7] SingleCellExperiment_1.21.1 fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [9] RCurl_1.98-1.12             lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt; [11] BiocFileCache_2.7.2         digest_0.6.31              </span></span>
<span><span class="co">#&gt; [13] lifecycle_1.0.3             cluster_2.1.4              </span></span>
<span><span class="co">#&gt; [15] RSQLite_2.3.1               magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [17] compiler_4.3.0              rlang_1.1.0                </span></span>
<span><span class="co">#&gt; [19] sass_0.4.5                  tools_4.3.0                </span></span>
<span><span class="co">#&gt; [21] igraph_1.4.2                utf8_1.2.3                 </span></span>
<span><span class="co">#&gt; [23] yaml_2.3.7                  knitr_1.42                 </span></span>
<span><span class="co">#&gt; [25] labeling_0.4.2              bit_4.0.5                  </span></span>
<span><span class="co">#&gt; [27] curl_5.0.0                  DelayedArray_0.25.0        </span></span>
<span><span class="co">#&gt; [29] withr_2.5.0                 purrr_1.0.1                </span></span>
<span><span class="co">#&gt; [31] desc_1.4.2                  grid_4.3.0                 </span></span>
<span><span class="co">#&gt; [33] fansi_1.0.4                 colorspace_2.1-0           </span></span>
<span><span class="co">#&gt; [35] scales_1.2.1                MASS_7.3-58.4              </span></span>
<span><span class="co">#&gt; [37] cli_3.6.1                   mzR_2.33.1                 </span></span>
<span><span class="co">#&gt; [39] rmarkdown_2.21              ragg_1.2.5                 </span></span>
<span><span class="co">#&gt; [41] generics_0.1.3              httr_1.4.5                 </span></span>
<span><span class="co">#&gt; [43] BiocBaseUtils_1.1.0         ncdf4_1.21                 </span></span>
<span><span class="co">#&gt; [45] DBI_1.1.3                   cachem_1.0.7               </span></span>
<span><span class="co">#&gt; [47] stringr_1.5.0               zlibbioc_1.45.0            </span></span>
<span><span class="co">#&gt; [49] parallel_4.3.0              AnnotationFilter_1.23.0    </span></span>
<span><span class="co">#&gt; [51] BiocManager_1.30.20         XVector_0.39.0             </span></span>
<span><span class="co">#&gt; [53] vctrs_0.6.1                 Matrix_1.5-4               </span></span>
<span><span class="co">#&gt; [55] jsonlite_1.8.4              bookdown_0.33              </span></span>
<span><span class="co">#&gt; [57] bit64_4.0.5                 clue_0.3-64                </span></span>
<span><span class="co">#&gt; [59] systemfonts_1.0.4           jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt; [61] glue_1.6.2                  pkgdown_2.0.7.9000         </span></span>
<span><span class="co">#&gt; [63] codetools_0.2-19            gtable_0.3.3               </span></span>
<span><span class="co">#&gt; [65] stringi_1.7.12              munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [67] tibble_3.2.1                pillar_1.9.0               </span></span>
<span><span class="co">#&gt; [69] htmltools_0.5.5             GenomeInfoDbData_1.2.10    </span></span>
<span><span class="co">#&gt; [71] R6_2.5.1                    dbplyr_2.3.2               </span></span>
<span><span class="co">#&gt; [73] textshaping_0.3.6           rprojroot_2.0.3            </span></span>
<span><span class="co">#&gt; [75] evaluate_0.20               lattice_0.21-8             </span></span>
<span><span class="co">#&gt; [77] highr_0.10                  memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [79] bslib_0.4.2                 Rcpp_1.0.10                </span></span>
<span><span class="co">#&gt; [81] xfun_0.38                   MsCoreUtils_1.11.6         </span></span>
<span><span class="co">#&gt; [83] fs_1.6.1                    pkgconfig_2.0.3</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
